draft-ietf-jsonpath-base.md

```diff
@@ -1072,22 +1076,58 @@ The following examples show the array slice selector in use by a child segment.
 
 ### Filter selector {#filter-selector}
 
+Filter selectors are used to iterate over the elements or members of
+structured values, i.e., JSON arrays and objects.
+The structured values are identified in the nodelist offered by the
+child or descendant segment the filter selector is used in.
+
+For each iteration (element/member), a logical expression is
+evaluated, the *filter expression*, which decides whether the node of
+the element/member is selected.
+(While a logical expression evaluates to what mathematically is a
+Boolean value, we use the term *logical* to maintain a distinction from
+the Boolean values that JSON can represent.)
+
+During the iteration process, the filter expression receives the node
+of each array element or value in an object member of the structured value being
+filtered; this element or member value is then known as the *current node*.
+
+The current node can be used as the start of one or more JSONPath
+queries used inside subexpressions of the filter expression, notated
+via the current-node-identifier `@`.
+Each JSONPath query can be used either for testing existence of a
+result of the query, for obtaining a specific JSON value resulting
+from that query that can then be used in a comparison, or as a
+*function argument*.
+
+Within the logical expression for a filter selector, function
+expressions can be used to operate on nodelists and values.
+The set of available functions is extensible, with a number of
+functions predefined, see {{fnex}}, and the ability to register further
+functions in the Function Extensions sub-registry ({{iana-fnex}}).
+When a function is defined, it is given a unique name, and the
+*declared type*
+of each of its arguments as well as the declared type returned from the
+function is declared.
+The type system is limited in scope; its function is to express
+restrictions that, without functions, are implicit in the grammar of
+filter expressions.
```

Comment:
Maybe this case can be added to the [conversion section](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/403/files#diff-aca3bb236d11f768483f8a1153cb9030b67d9cd6a39f012e8e4bfec2c6275165R1538) as another "gotcha" example.

While I agree that it's surprising, the spec doesn't claim to have a way to compare nodelists, and I think, once thought through, a result of `true` for your example can be expected.

Reply:
I'm actually leaning in the inverse direction: getting rid of "singular types" in the grammar.
But it would be good to have a proposal for going in the direction you suggest.
Not in this PR, though...

Reply:
To be clear, what do you mean by "singular types?"

Reply:
`singular-path` as a feeble duplication of `filter-path`.
(We also need to fix the suffix "path" in these things.  Later...)